tools/backups2datalad/zarr.py

```diff
@@ -0,0 +1,405 @@
+from urllib.parse import urlparse, urlunparse
+
+import boto3
+from botocore import UNSIGNED
+from botocore.client import Config as BotoConfig
+from dandi.dandiapi import RemoteZarrAsset, RemoteZarrEntry, ZarrEntryStat, ZarrListing
+from datalad.api import Dataset
+import dateutil.parser
+import httpx
+import trio
+
+from .annex import AsyncAnnex
+from .consts import ZARRS_REMOTE_PREFIX, ZARRS_REMOTE_UUID
+from .util import (
+    Config,
+    MiniFuture,
+    Remote,
+    arequest,
+    create_github_sibling,
+    custom_commit_date,
+    init_dataset,
+    key2hash,
+    log,
+    maxdatetime,
+    quantify,
+)
+
+if sys.version_info[:2] >= (3, 10):
+    from contextlib import aclosing
+else:
+    from async_generator import aclosing
+
+
+CHECKSUM_FILE = ".zarr-checksum"
+
+
+@dataclass
+class ZarrReport:
+    added: int = 0
+    updated: int = 0
+    deleted: int = 0
+    checksum: bool = False
+
+    def __bool__(self) -> bool:
+        return bool(self.added or self.updated or self.deleted or self.checksum)
+
+    def get_summary(self) -> str:
+        msgparts = []
+        if self.added:
+            msgparts.append(f"{quantify(self.added, 'file')} added")
+        if self.updated:
+            msgparts.append(f"{quantify(self.updated, 'file')} updated")
+        if self.deleted:
+            msgparts.append(f"{quantify(self.deleted, 'file')} deleted")
+        if self.checksum:
+            msgparts.append("checksum updated")
+        if not msgparts:
+            msgparts.append("No changes")
+        return ", ".join(msgparts)
+
+
+@dataclass
+class ZarrSyncer:
+    api_url: str
+    zarr_id: str
+    repo: Path
+    annex: AsyncAnnex
+    s3client: httpx.AsyncClient
+    backup_remote: Optional[str]
+    checksum: str
+    last_timestamp: Optional[datetime] = None
+    extant_paths: Set[str] = field(default_factory=set)
+    report: ZarrReport = field(default_factory=ZarrReport)
+
+    async def run(self) -> None:
+        async with aclosing(self.aiter_file_entries()) as ait:  # type: ignore[type-var]
+            async for entry in ait:
+                log.info("Zarr %s: %s: Syncing", self.zarr_id, entry)
+                self.extant_paths.add(str(entry))
+                bucket_url, st = await self.stat(entry)
+                md5_digest = entry.get_digest().value
+                self.last_timestamp = maxdatetime(self.last_timestamp, st.modified)
+                dest = self.repo / str(entry)
+                if dest.is_dir():
+                    log.info(
+                        "Zarr %s: %s: deleting conflicting directory at same path",
+                        self.zarr_id,
+                        entry,
+                    )
+                    await trio.to_thread.run_sync(self.rmtree, dest)
+                else:
+                    for ep in entry.parents:
+                        pp = self.repo / str(ep)
+                        if pp.is_file() or pp.is_symlink():
```

Comment:
how could we ended up with a symlink here? or could that symlink be a symlink to a folder, thus may be 'legit' and not worth removing?

Reply:
If a Zarr originally contains a file at `foo/bar`, and then it's modified so that `foo/bar` becomes a directory containing a file `baz`, we need to delete the old `foo/bar` file when updating the Zarr backup.  `pp.is_symlink()` catches broken symlinks, which can occur when the file content is not present in the local annex.

Reply:
RRIGHT! it is all annexed so those would not be files but symlinks (broken or not should not matter)!
```suggestion
                        # git/annexed content could be a file or a symlink but
                        # now should become a directory
                        if pp.is_file() or pp.is_symlink():
```
not sure if indentation didn't do some space/TABs conversions since diff looks a bit off (I just added lines)

Reply:
I added comments in d6e21c0.